File: src/main/java/com/machfour/macros/cli/Read.java
package com.machfour.macros.cli;

import com.machfour.macros.linux.Config;
import com.machfour.macros.linux.LinuxDatabase;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDataSource;
import com.machfour.macros.storage.MacrosDatabase;
import org.jetbrains.annotations.NotNull;

import java.io.FileReader;
import java.io.IOException;
import java.io.PrintStream;
import java.io.Reader;
import java.sql.SQLException;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.cli.CliMain.OUT;
import static com.machfour.macros.cli.CliMain.PROGNAME;
import static com.machfour.macros.cli.CliMain.ERR;


class Read extends CommandImpl {
    private static final String NAME = "read";
    private static final String USAGE = String.format("Usage: %s %s <file> [-v | --verbose] [--per100]", PROGNAME, NAME);

    Read() {
        super(NAME, USAGE);
    }

    @Override
    public void doAction(List<String> args) {
        if (args.size() < 2) {
            printHelp();
            out.println();
            out.println("Please specify a file to read");
            return;
        } else if (args.contains("--help")) {
            printHelp();
            return;
        }
        String filename = args.get(1);
        boolean verbose = args.contains("--verbose") || args.contains("-v");
        boolean per100 = args.contains("--per100");

        MacrosDatabase db = LinuxDatabase.getInstance(Config.DB_LOCATION);

        FileParser fileParser = new FileParser();
        List<Meal> meals;
        try (Reader r = new FileReader(filename)) {
            meals = fileParser.parseFile(r, db);
        } catch (IOException e1) {
            ERR.println("IO exception occurred: " + e1.getMessage());
            return;
        } catch (SQLException e2) {
            ERR.println("SQL exception occurred: " + e2.getMessage());
            return;
        }
        MealPrinter.printMeals(meals, OUT, verbose, per100, true);
        Map<String, String> errors = fileParser.getErrorLines();
        if (!errors.isEmpty()) {
            out.println();
            out.println("Warning: the following lines were skipped");
            for (Map.Entry<String, String> line : errors.entrySet()) {
                out.printf("'%s' - %s\n", line.getKey(), line.getValue());
            }
        }
    }
}

File: macros-cli/src/main/kotlin/com/machfour/macros/cli/modes/Read.kt
package com.machfour.macros.cli.modes

import com.machfour.macros.cli.CliConfig
import com.machfour.macros.cli.CommandImpl
import com.machfour.macros.cli.utils.ArgParsingResult
import com.machfour.macros.cli.utils.findArgumentFromFlag
import com.machfour.macros.cli.utils.printMeals
import com.machfour.macros.cli.utils.printlnErr
import com.machfour.macros.entities.Meal
import com.machfour.macros.insulin.BolusCalculator
import com.machfour.macros.insulin.parseInsulinArgument
import com.machfour.macros.insulin.printInsulin
import com.machfour.macros.parsing.FileParser
import com.machfour.macros.sql.SqlException
import java.io.FileReader
import java.io.IOException


class Read(config: CliConfig) : CommandImpl(config) {
    override val name = "read"
    override val usage = "Usage: ${config.programName} $name <file> " +
            "[ ${verboseFlag.usage} ] [${per100Flag.usage}] [${insulinFlag.full} <ic ratio>[:<protein factor]]"

    data class Flag(
        val name: String,
        val mnemonic: String? = null,
        val hasArg: Boolean = false,
        val argOptional: Boolean = false
    ) {
        val full: String
            get() = "--$name"
        val abbr: String?
            get() = mnemonic?.let { "-$it" }
        val usage: String
            get() = (if (mnemonic != null) "$abbr | " else "") + full

        fun containedIn(cmdline: List<String>) : Boolean {
            return cmdline.contains(full) || cmdline.contains(abbr)
        }
    }

    companion object {
        private val helpFlag = Flag("help", "h")
        private val verboseFlag = Flag("verbose", "v")
        private val per100Flag = Flag("per100")
        private val insulinFlag = Flag("insulin")
    }

    override fun doAction(args: List<String>): Int {
        if (args.size < 2) {
            printHelp()
            println()
            println("Please specify a file to read")
            return 2
        } else if (helpFlag.containedIn(args)) {
            printHelp()
            return 2
        }
        val filename = args[1]
        val verbose = verboseFlag.containedIn(args)
        val per100 = per100Flag.containedIn(args)

        val insulinArg = findArgumentFromFlag(args, insulinFlag.full)
        val insulinCalc: BolusCalculator?

        when (insulinArg) {
            is ArgParsingResult.KeyValFound -> {
                insulinCalc = parseInsulinArgument(insulinArg.argument)
                if (insulinCalc == null) {
                    printlnErr(
                        "I:C ratio and fat/protein factors (if used) in ${insulinFlag.full} " +
                                " argument must be numeric, positive and separated by ':' characters"
                    )
                    return 1
                }
            }
            is ArgParsingResult.ValNotFound -> {
                printlnErr("${insulinFlag.full} requires an argument")
                return 1
            }
            else -> {
                insulinCalc = null
            }
        }

        val fileParser = FileParser()
        val meals: List<Meal>
        try {
            FileReader(filename).use { meals = fileParser.parseFile(config.database, it.readLines()) }
        } catch (e1: IOException) {
            printlnErr("IO exception occurred: " + e1.message)
            return 1
        } catch (e2: SqlException) {
            printlnErr("SQL exception occurred: " + e2.message)
            return 1
        }

        printMeals(meals, verbose, per100, true)
        val errors = fileParser.getErrorLines()
        if (errors.isNotEmpty()) {
            println()
            println("Warning: the following lines were skipped")
            for ((key, value) in errors) {
                println("'$key' - $value")
            }
            return 2
        }

        if (insulinCalc != null) {
            printInsulin(Meal.sumNutrientData(meals), insulinCalc)
        }

        return 0
    }
}

